# 4 Deriving New Variables

## Table of contents
1. [Importing Libraries & Data](#imports)
2. [Price Labels](#paragraph1)
3. [Busyness label](#paragraph2)
4. [Busiest days label](#paragraph3)
5. [Busiest period of day label](#paragraph4)
6. [Data export](#paragraph5)

### 1. Importing Libraries & Data <a name="imports"></a>

In [1]:
# Importing pandas, numpy and os libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Turning project folder path into a string
path = r'C:\Users\deuts\Desktop\Instacart Basket Analysis'

# Importing 'orders_products_merged.pkl'
orders_products_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [3]:
orders_products_merged.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0 product_name  aisle_id  department_id  prices _merge  
0         195         Soda        77              7     9.0   both  
1         195         Soda        77              7     9.0   both  
2         195         Soda        77              7     9.0   both  
3         195         Soda        77              7     9.0   both  
4         195         Soda        77              7     9.0   both

In [4]:
df = orders_products_merged[:1000000]

In [5]:
df.shape

(1000000, 15)

### 2. Price labels <a name="paragraph1"></a>

In [6]:
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [7]:
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\deuts\AppData\Local\Temp\ipykernel_2000\2848743372.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [8]:
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [9]:
df['prices'].max()

14.8

In [10]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\deuts\AppData\Local\Temp\ipykernel_2000\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [11]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [12]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [13]:
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

In [14]:
orders_products_merged.loc[orders_products_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [15]:
orders_products_merged.loc[(orders_products_merged['prices'] <= 15) & (orders_products_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [16]:
orders_products_merged.loc[orders_products_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
orders_products_merged['price_range_loc'].value_counts(dropna = True)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

In [18]:
# Renaming column from 'order_dow' to 'orders_day_of_the_week' again since I didnt't save it properly apperantly 
orders_products_merged = orders_products_merged.rename(columns={'order_dow': 'orders_day_of_week'})
orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0 product_name  aisle_id  department_id  prices _merge  \
0         195         Soda        77              7     9.0   both   
1         195         Soda        77              7     9.0   both   
2         195         Soda        77              7     9.0   both   
3         195         Soda        77              7     9.0   both   
4         195         Soda        77              7     9.0   both   

     price_range_loc  
0  Mid-range product  
1  Mid-range product  
2  Mid-range product  
3  Mid-range product  
4  Mid-range product

### 3. Busyness label <a name="paragraph2"></a>

In [19]:
orders_products_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [20]:
result = []

for value in orders_products_merged['orders_day_of_week']:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [21]:
orders_products_merged['busiest_day'] = result

In [22]:
orders_products_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

### 4. Busiest days label <a name="paragraph3"></a>

In [23]:
## 2. Busiest days
# Creating for loop for checking the busiest days of the week
result_2 = []

for value in orders_products_merged['orders_day_of_week']:
  if value in [0,1]:
    result_2.append("Busiest day")
  elif value in [3,4]:
    result_2.append("Least busy")
  else:
    result_2.append("Regularly busy")

In [24]:
# Creating new column 'busiest_days' with the results of the for loop
orders_products_merged['busiest_days'] = result_2
orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0 product_name  aisle_id  department_id  prices _merge  \
0         195         Soda        77              7     9.0   both   
1         195         Soda        77              7     9.0   both   
2         195         Soda        77              7     9.0   both   
3         195         Soda        77              7     9.0   both   
4         195         Soda        77              7     9.0   both   

     price_range_loc     busiest_day    busiest_days  
0  Mid-range product  Regularly busy  Regularly busy  
1  Mid-range product  Regularly busy      Least busy  
2  Mid-range product  Regularly busy      Least busy  
3  Mid-range product      Least busy      Least busy  
4  Mid-range product      Least busy      Least busy

In [25]:
# Counting values of new column, sorted by business of days
orders_products_merged['busiest_days'].value_counts(dropna = False)

Regularly busy    12916111
Busiest day       11864412
Least busy         7624336
Name: busiest_days, dtype: int64

In [26]:
# 3. Checking result against output of business per day by the sum of the according single days
orders_products_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

###### The sum of the days 0 (Saturday) & 1 (sunday) as beeing the busiest days is 6204182 + 5660230 = 11864412 equals the same as in the displayed data per day.
###### The sum of the days 3 (tuesday) & 4 (Wednesday) as beeing the slowest days is 3840534 + 3783802 = 7624336 equals the same as in the displayed data data per day.

### 5. Busiest period of day label <a name="paragraph4"></a>

In [27]:
## 4. Busiest hours
# Displaying orders per hour descending by orders
orders_products_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [28]:
# Creating for loop for sorting the busiest hours of the day by periods
result_3 = []

for value in orders_products_merged['order_hour_of_day']:
  if value>=9 and value<=16:
    result_3.append("Most orders")
  elif value<=6:
    result_3.append("Fewest orders")
  else:
    result_3.append("Average orders")

In [29]:
# Creating new column 'busiest_period_of_day' with the results of the for loop
orders_products_merged['busiest_period_of_day'] = result_3
orders_products_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

   Unnamed: 0 product_name  aisle_id  department_id  prices _merge  \
0         195         Soda        77              7     9.0   both   
1         195         Soda        77              7     9.0   both   
2         195         Soda        77              7     9.0   both   
3         195         Soda        77              7     9.0   both   
4         195         Soda        77              7     9.0   both   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regularly busy        Average orders  
1  Mid-range product  Regularly busy      Least busy        Average orders  
2  Mid-range product  Regularly busy      Least busy           Most orders  
3  Mid-range product      Least busy      Least busy        Average orders  
4  Mid-range product      Least busy      Least busy           Most orders

In [30]:
# Counting values of new column by determined periods
orders_products_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders    10399967
Fewest orders       886821
Name: busiest_period_of_day, dtype: int64

### 6. Data export <a name="paragraph5"></a>

In [ ]:
## 7. Export
# Exporting data frame as pickle file to prepared data folder
orders_products_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated.pkl'))